File: src/SIL.Machine.TS/src/tokenization/string-tokenizer.ts
import { createRange, Range } from '../annotations/range';
import { Tokenizer } from './tokenizer';

export abstract class StringTokenizer implements Tokenizer {
  abstract tokenize(data: string, range?: Range): Range[];

  tokenizeToStrings(data: string, range: Range = createRange(0, data.length)): string[] {
    return this.tokenize(data, range).map(r => data.substring(r.start, r.end));
  }
}

File: src/SIL.Machine/Tokenization/StringTokenizer.cs
using System.Collections.Generic;
using System.Linq;
using SIL.Machine.Annotations;

namespace SIL.Machine.Tokenization
{
    public abstract class StringTokenizer : IRangeTokenizer<string, int, string>
    {
        public IEnumerable<string> Tokenize(string data)
        {
            return Tokenize(data, Range<int>.Create(0, data.Length));
        }

        public IEnumerable<string> Tokenize(string data, Range<int> range)
        {
            return TokenizeAsRanges(data, range).Select(r => data.Substring(r.Start, r.Length));
        }

        public IEnumerable<Range<int>> TokenizeAsRanges(string data)
        {
            return TokenizeAsRanges(data, Range<int>.Create(0, data.Length));
        }

        public abstract IEnumerable<Range<int>> TokenizeAsRanges(string data, Range<int> range);
    }
}
